<a href="https://colab.research.google.com/github/Jast15/data-analysis/blob/main/%D0%97%D0%B0%D0%BD%D1%8F%D1%82%D0%B8%D0%B5_4_(2).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Анализ и обработка результатов

В анализе данных важно наглядное представление результатов, чтобы их мог оценить заказчик. Это бизнес, где на кону громадный трафик, серверные мощности, личное время многотысячной аудитории. И одновременно это исследование.

Как в настоящей экспериментальной науке, ответ непредсказуем. Его точность зависит от умелого владения статистическими методами и от качества исходных данных. Перед тем, как начинать считать, проверьте, грамотно ли ваши данные подготовлены.

# Группировка данных

Слово «анализ» означает разбор, рассмотрение с разных сторон. Анализ данных начинают с разделения их на группы по какому-нибудь признаку. Эта операция называется группировка данных. Она помогает изучить материал более подробно, чтобы затем перейти к поиску взаимосвязей между отдельными группами.


Группировка оправданна, если данные чётко делятся по значимому признаку, а полученные группы близки к теме задачи. Например, когда есть данные обо всех покупках в супермаркете, можно смело заниматься группировкой. Так можно установить время наплыва покупателей и решить проблему пиковых нагрузок. Или посчитать средний чек — обычно для магазинов это ключевая метрика.


Стадии группировки хорошо укладываются в словесную формулу **split-apply-combine**:

•	разделить, **split** — разбиение на группы по определённому критерию;

•	применить, **apply** — применение какого-либо метода к каждой группе в отдельности, например, подсчёт численности группы методом count() или суммирование вызовом sum();

•	объединить, **combine** — сведение результатов в новую структуру данных, в зависимости от условий разделения и выполнения метода это бывает DataFrame и Series.

На картинке изображен принцип **split-apply-combine** для таблицы с экзопланетами. Посмотрим, как вообще идут дела с поиском экзопланет. Сначала данные делят по группам, где каждая группа — это год. Потом метод **count()** подсчитывает численность каждой группы. В итоге получаем новую структуру данных с группами, где каждая содержит год и число открытых за этот год экзопланет.


В Рandas для группировки данных есть метод **groupby()**. Он принимает как аргумент название столбца, по которому нужно группировать. В случае с делением экзопланет по годам открытия:

**print(exoplanet.groupby('discovered'))**

**<pandas.core.groupby.DataFrameGroupBy object at 0x7fc1e1ca3400>**


Применение метода **groupby()** к объекту типа DataFrame приводит к созданию объекта особого типа — **DataFrameGroupBy**. Это сгруппированные данные. Если применить к ним какой-нибудь метод Pandas, они станут новой структурой данных типа **DataFrame** или **Series**.
Подсчитаем сгруппированные по годам экзопланеты методом **count()**:

Если нужно сравнить наблюдения по одному показателю, метод применяют к **DataFrameGroupBy** с указанием на один столбец. Нас в первую очередь интересует радиус экзопланет: мы ищем другую Землю. Давайте получим таблицу с единственным столбцом 'radius':

**exo_number = exoplanet.groupby('discovered')['radius'].count()**

**print(exo_number)**

Получили Series, где по годам открытия расписано количество экзопланет, для которых удалось установить радиус.

Посмотрим, как меняется средний радиус открытых экзопланет год от года. Для этого надо сложить радиусы планет, открытых за определённый год, и поделить на их количество (которое мы уже нашли).

Сумма радиусов считается методом **sum()**:

**exo_radius_sum = exoplanet.groupby('discovered')['radius'].sum()**


**print(exo_radius_sum)**



## Задача
Вашему вниманию предлагается датасет с обзором показателей самоубийств с 1985 по 2016 год — Сравнение социально-экономической информации с показателями самоубийств по годам и странам.

Столбцами этой таблицы являются: страна, год, пол, возрастная группа, количество самоубийств, население, уровень самоубийств, сводный ключ по стране за год, ИЧР (индекс человеческого развития) за год, ВВП за год, ВВП на душу населения, поколение (на основе среднего по возрастной группе).

1. Загрузите таблицу master, содержащую данные о самоубийствах. Выведите на экран таблицу и оцените данные, обратите внимание на названия столбцов, переименуйте их, если это необходимо.

2. Убедитесь в том, что данные прошли предподготовку.  Пропущенные и неопределённые значения выявляет метод isna(), а суммарное количество таких значений — метод sum().

3. Сгруппируйте DataFrame по столбцу country, сохраните полученный результат в переменной country_grouping.

4. Посчитайте количество случаев самоубийств по странам, которые были совершены, методом **sum()**, указав, что выбираем один столбец suicides_no.

Сохраните результат в переменной country_sum и выведите первые 30 строк этой таблицы.



In [ ]:
import pandas as pd

df = pd.read_csv('master.csv')

print(df.head())

print("Названия столбцов:", df.columns.tolist())

# Проверка наличия пропущенных значений
missing_values_before = df.isna().sum()
print("Количество пропущенных значений по столбцам перед удалением:\n", missing_values_before)

# Удаление строк с пропущенными значениями
df_cleaned = df.dropna()

# Проверка наличия пропущенных значений после удаления
missing_values_after = df_cleaned.isna().sum()
print("Количество пропущенных значений по столбцам после удаления:\n", missing_values_after)

# Группировка по стране
country_grouping = df.groupby('country')

# Подсчет общего количества самоубийств по странам
country_sum = country_grouping['suicides_no'].sum().reset_index()

# Вывод первых 30 строк таблицы с количеством самоубийств по странам
print(country_sum.head(30))





   country  year     sex          age  suicides_no  population  \
0  Albania  1987    male  15-24 years           21      312900   
1  Albania  1987    male  35-54 years           16      308000   
2  Albania  1987  female  15-24 years           14      289700   
3  Albania  1987    male    75+ years            1       21800   
4  Albania  1987    male  25-34 years            9      274300   

   suicides/100k pop country-year  HDI for year  gdp_for_year ($)   \
0               6.71  Albania1987           NaN      2,156,624,900   
1               5.19  Albania1987           NaN      2,156,624,900   
2               4.83  Albania1987           NaN      2,156,624,900   
3               4.59  Albania1987           NaN      2,156,624,900   
4               3.28  Albania1987           NaN      2,156,624,900   

   gdp_per_capita ($)       generation  
0                 796     Generation X  
1                 796           Silent  
2                 796     Generation X  
3                 

# Сортировка данных


Поиск необычного в группе — что среди планет, что среди меломанов — это прежде всего поиск чемпионов: объектов с выдающимися показателями по разным статьям. Как всю таблицу, так и отдельные группы изучают, сортируя строки по какому-либо столбцу.

В Pandas для этой операции есть метод **sort_values()**. У него два аргумента:

• **by = 'имя столбца'** — имя столбца, по которому нужно сортировать;

• **ascending:** по умолчанию True. Для сортировки по убыванию установите значение False.



Среди экзопланет интересны близкие по размерам к Земле. Есть ли такие? Отсортируем список по радиусу в порядке возрастания. Тогда в голове таблицы окажутся самые малые, на которых гравитация не прижмёт нас к полу.

**print(exoplanet.sort_values(by='radius').head(30))**

Оказывается, некоторые из уже открытых экзопланет по размерам близки не то что к Земле, но уже и к Луне! Получим список экзопланет с радиусом меньше земного. Смотрите, как логический оператор (здесь это <) задействуется в отборе элементов столбца. Дальше нам этот приём не раз понадобится.

**print(exoplanet[exoplanet['radius'] < 1])**

Но и этот список такой длинный, что изучать его лучше по частям. Экзопланеты, близкие по размерам к Земле, за последнее десятилетие открывали нередко. Можно изучать список открытых за каждый год. Например, для 2014 года (вновь обратите внимание на работу логического оператора, теперь это ==):

**print(exoplanet[exoplanet['discovered'] == 2014])**

А чтобы не тратить время на лишнее, поставим оба условия сразу. Для этого в Pandas есть логический оператор &, подобный оператору and языка Python. Напомним, его смысл на русском языке можно передать словами «и ещё»:

**exo_small_14 = exoplanet[ (exoplanet['radius']<1) & (exoplanet['discovered']==2014)]**

**print(exo_small_14)**

Отсортируем результат в порядке убывания радиуса.

**print(exo_small_14.sort_values(by = 'radius', ascending = False))**





## Задача
1. Выделим пятерки стран с самым низким и самым высоким уровнем самоубийств в мире.

In [ ]:
lowest_suicides = country_sum.nsmallest(5, 'suicides_no')
highest_suicides = country_sum.nlargest(5, 'suicides_no')

print("Пять стран с самым низким количеством самоубийств:")
print(lowest_suicides[['country', 'suicides_no']])

print("\nПять стран с самым высоким количеством самоубийств:")
print(highest_suicides[['country', 'suicides_no']])


Пять стран с самым низким количеством самоубийств:
                  country  suicides_no
27               Dominica            0
76  Saint Kitts and Nevis            0
79             San Marino            4
1     Antigua and Barbuda           11
55               Maldives           20

Пять стран с самым высоким количеством самоубийств:
               country  suicides_no
75  Russian Federation      1209742
98       United States      1034013
46               Japan       806902
33              France       329127
95             Ukraine       319950


2. Исследуйте распределение количества суидидов по гендерному признаку. Найдите лидеров и антилидеров по количеству самоубийств среди женщин. Повторяется ли картина по сравнению с предыдущим исследованием?

In [ ]:
women_suicides = df_cleaned[df_cleaned['sex'] == 'female']

country_women_suicides = women_suicides.groupby('country').agg({'suicides_no': 'sum'}).reset_index()

lowest_women_suicides = country_women_suicides.nsmallest(1, 'suicides_no')
highest_women_suicides = country_women_suicides.nlargest(1, 'suicides_no')

print("Страна с самым низким количеством самоубийств среди женщин:")
print(lowest_women_suicides[['country', 'suicides_no']])

print("\nСтрана с самым высоким количеством самоубийств среди женщин:")
print(highest_women_suicides[['country', 'suicides_no']])

print("Вывод: по сравнению с предыдущим иследованием данные немного различаются,\nно страны с самым низким и высоким количеством самоубийств среди женщин \nвходит в топ 5, предыдущего иследования.")


Страна с самым низким количеством самоубийств среди женщин:
               country  suicides_no
1  Antigua and Barbuda            0

Страна с самым высоким количеством самоубийств среди женщин:
   country  suicides_no
43   Japan        80683
Вывод: по сравнению с предыдущим иследованием данные немного различаются,
но страны с самым низким и высоким количеством самоубийств среди женщин 
входит в топ 5, предыдущего иследования.


# Описательная статистика

Чтобы представить себе среднестатистического самоубийцу, понадобится описательная статистика.

Из её показателей для количественного описания данных в нашей задаче нужны четыре меры — максимум, минимум, медиана и среднее.

Наибольшее и наименьшее обычно вычисляют только по одному признаку. Например, можно получить минимальное и максимальное значение уровня самоубийств (количество самоубийств на 100 тыс.населения) (столбец 'suicides/100k pop'). Для поиска максимума вызывают метод max(), примененный к этому столбцу:

print(master['suicides/100k pop'].max())


In [ ]:
print(df['suicides/100k pop'].max())

224.97


Самый высокий показатель оказался 224.97 самоубийств на 100 тыс.населения. Интересно, в какой стране такая ситуация.

Запросим из master строку с максимальным значением, прибегнув к логической индексации с условием

**master['suicides/100k pop'] == master['suicides/100k pop'].max()**

In [ ]:
max_suicide_rate = df['suicides/100k pop'].max()
country_max = df[df['suicides/100k pop'] == max_suicide_rate]
print("Страна с максимальным уровнем самоубийств:")
print(country_max[['country', 'suicides/100k pop']])

Страна с максимальным уровнем самоубийств:
     country  suicides/100k pop
1258   Aruba             224.97


Минимальное значение — ищут методом **min()**. Нам интересно установить случаи, когда уровень суицидов минимальный, но не нулевой.

Вот как мы создаём выборку  и находим в ней минимальное значение:

**master_drop_null = master[master['suicides/100k pop'] != 0]**

**print(master_drop_null['suicides/100k pop'].min())**


In [ ]:
master_drop_null = df[df['suicides/100k pop'] != 0]

min_suicide_rate = master_drop_null['suicides/100k pop'].min()

country_min = df[df['suicides/100k pop'] == min_suicide_rate]
print("Страна с минимальным уровнем самоубийств (не нулевой):")

print(country_min[['country', 'suicides/100k pop']])

Страна с минимальным уровнем самоубийств (не нулевой):
            country  suicides/100k pop
23347  South Africa               0.02
23371  South Africa               0.02
23479  South Africa               0.02
25060      Thailand               0.02
25061      Thailand               0.02


Получим название самой "счастливой" страны. Как и при поиске максимума, воспользуемся условием и логической индексацией

Результат вывода — две страны, которые соответствуют минимальному значению.
На основе полученных данных можно сделать вывод, что число совершенных самоубийств на 100 тыс. населения  находится в диапазоне от 0,02 до 224.97 секунд, не включая пропущенные.

Это знание пригодится, чтобы разобраться с медианой и средним арифметическим.

Среднее и медиана оценивают значения в центре выборки. Если тех строк, где число с большим количеством самоубийц, столько же, сколько тех, где это число мало — среднее подойдет. Но когда есть оторванные от основной массы лидеры, их результаты сильно смещают значение среднего вверх. Вот почему оценивать в этом случае лучше медианой.



В геометрии медиана делит фигуру на две равные по площади. В статистике она делит выборку пополам: в одной половине значения меньше медианного, в другой больше. Логично, что для определения медианы список обязательно должен быть отсортирован — либо по возрастанию, либо по убыванию.
Когда количество значений нечётное, медиана будет равна тому значению, которое оказалось ровно посередине отсортированного набора. Если же количество данных чётное, то медиана рассчитывается как среднее арифметическое двух соседних чисел в середине набора.



В Pandas есть метод **median()**, который считает медиану. По аналогии с **min()** и **max()** его можно применять ко всей таблице, к отдельному столбцу или к сгруппированным данным. Теперь применим медиану ко всем значениям времени прослушивания в нашей таблице, исключив нулевые:



In [ ]:
df_filtered = df[df['suicides/100k pop'] != 0]

median_suicides = df_filtered['suicides/100k pop'].median()

print(f"Медиана уровня самоубийств (исключая нулевые значения): {median_suicides}")

Медиана уровня самоубийств (исключая нулевые значения): 8.26


Чтобы убедиться, что лидеры действительно смещают средний показатель вверх, найдём среднее арифметическое всех этих значений методом mean():

In [ ]:
suicides = df['suicides/100k pop'].mean()
print(f"Среднее значение уровня самоубийств: {suicides}")

Среднее значение уровня самоубийств: 12.816097411933864


In [7]:
import pandas as pd

df = pd.read_csv('master.csv')

italy_men_suicides = df[(df['country'] == 'Italy') &
                         (df['year'].between(2001, 2003)) &
                         (df['sex'] == 'male')]

total_men_suicides = italy_men_suicides['suicides_no'].sum()

print(f"Количество мужчин, покончивших жизнь самоубийством в Италии с 2001 по 2003: {total_men_suicides}")




Количество мужчин, покончивших жизнь самоубийством в Италии с 2001 по 2003: 9273
